pkg/cri/server/image_pull.go

```diff
@@ -92,7 +93,18 @@ import (
 // contents are missing but snapshots are ready, is the image still "READY"?
 
 // PullImage pulls an image with authentication config.
-func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (*runtime.PullImageResponse, error) {
+func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (_ *runtime.PullImageResponse, err error) {
+	var domain string
+	defer func() {
+		if err != nil {
+			imagePulls.WithValues("failed", domain).Inc()
+		} else {
+			imagePulls.WithValues("succeed", domain).Inc()
+		}
+	}()
```

Comment:
Bringing up https://github.com/containerd/containerd/pull/7313#pullrequestreview-1124091873 again: I don't think it's reasonable to assume that containerd will only pull from 10 (or 100) domains within its lifetime.  Some container registry providers use many distinct domains and some users of containerd may not be controlling the registries where images are pulled from.

If we want to produce labeled metrics that distinguish image sources, we may need to provide a mechanism for the user to configure the set of registries that they are interested in (or potentially the set of images they're interested in too).  Or we could allow that information to be read and post-processed from the containerd log given the debug output on line 209 instead of exposing it in Prometheus.

Reply:
To make this forward, I think we may remove the domain label here in this PR.
- TODO: `configure the set of registries that they are interested in` may be another feature based on this PR.

Reply:
@samuelkarp we have use cases to measure latency for each image (label). For common internal K8s or other docker based deployments, number of container registries of an organization usually tend to be limited (etc. one per regions), this feels like a common case to me unless it's for cloud services such as ECS, Fargate, Lambda etc.

However to label by image, the cardinality could be an issue for Prometheus, but many organizations could make their Prom scalable, so it seems to be reasonable to leave per-image metric open for opting in.

That said, I like the idea of using service metrics logs instead of trying to Figure out Prom metrics (debating over gauge, histogram, summary for these fun stuff, deciding/configuring what dimension to label, thinking about cardinality etc), I feel it will make containerd more complex than it should be.

Adding too many Prom metrics into containerd also makes it easy to break previous metrics, hard to test, and slower to iterate on. 

Using CRI.PullImage as an example, if we can produce the following structured log, it's up to the containerd user to parse the logs, decide what dimensions to expose, having a daemonset to export the metrics based on their need. Some rough idea:

{
"operation": "CRI.PullImage",
"latencyMs": 12,
"imageRef": "ubuntu:latest",
"imageSizeMB": 65,
"downloadBlobsLatencyMs": 2000,
"unpackLatencyMs": 4000  
}

^ idea is also applicable for other pod/container CRUD APIs

Let me know what you think, happy to discuss since it seems to be a very common ask.

Reply:
I would prefer to add further flags to enable such a feature.

As @samuelkarp  mentioned, we should not use the domain metrics label by default because there are some cases that it will make the metrics too large. Using `imageRef` will make it even larger.

However, I am open to adding a further flag like `--enable-pull-metrics-per-image` or `--enable-pull-metrics-per-domain` to make it show more details. (In my case, a cluster with several local registries/harbors, we would prefer to add a registry label if it can.) 

For short, the current PR can provide general image pull metrics. This would be a good start IMO.
We can work on more detailed image-pulling metrics after we get some feedback. (Maybe we can open an issue/discussion to discuss about it or collect feedbacks around metrics)

Reply:
We discussed a related issue in kubelet/sig-node in this week's weekly meeting. See https://docs.google.com/document/d/1Ne57gvidMEWXR70OxxnRkYquAoMpt56o75oZtg-OeBg/edit#heading=h.6kx17lyebeal. 

To me, the most important metrics for image pulling is the current parallel pulling count `inProgressImagePulls`.    This would help us know how many image-pullings are in process, and it may help us to set a better limitation on the kubelet side to limit the image pull QPS/burst(current flags). This is just a start in this area. We can move forward one step by another.